src/plugins/validation/2and3/semantic-validators/items-required-for-array-objects.js

```diff
@@ -10,8 +10,47 @@
 
 const { walk } = require('../../../utils');
 const MessageCarrier = require('../../../utils/messageCarrier');
+const at = require('lodash/at');
 
-module.exports.validate = function({ jsSpec }) {
+const reduceObj = function(jsSpec, obj) {
+  if (obj['$ref']) {
+    const objPath = obj['$ref'].split('/');
+    objPath.shift();
+    return reduceObj(jsSpec, at(jsSpec, [objPath])[0]);
+  }
+  return obj;
+};
+
+const checkReqProp = function(jsSpec, obj, requiredProp) {
+  obj = reduceObj(jsSpec, obj);
+  if (obj.properties && obj.properties[requiredProp]) {
+    return true;
+  } else if (Array.isArray(obj.anyOf) || Array.isArray(obj.oneOf)) {
+    const childList = obj.anyOf || obj.oneOf;
+    let reqPropDefined = true;
+    childList.forEach((childObj) => {
+      if (childObj.properties && !childObj.properties[requiredProp]) {
+        reqPropDefined = false;
+      } else {
+        reqPropDefined = reqPropDefined && checkReqProp(jsSpec, childObj, requiredProp);
+      }
+    });
+    return reqPropDefined;
+  } else if (Array.isArray(obj.allOf)) {
+    let reqPropDefined = false;
+    obj.allOf.forEach((childObj) => {
+      if (childObj.properties && childObj.properties[requiredProp]) {
+        reqPropDefined = true;
+      } else if (checkReqProp(jsSpec, childObj, requiredProp)) {
+        reqPropDefined = true;
+      }
+    });
```

Comment:
Just like above, I think this can be simplified:

```suggestion
    obj.allOf.forEach((childObj) => {
      if (checkReqProp(jsSpec, childObj, requiredProp)) {
        reqPropDefined = true;
      }
    });
```